### ISSUE TITLE
Blazor server dispose of javascript addEventListener does not work on page refresh

### ISSUE THREAD

**Issue by ScottyBlazor** [dotnet/aspnetcore#29747]
I have designed a blazor server component
in the OnAfterRender method I call a Javascript function to addEventListener to listen to keydown events on a textbox

I have implemented the IDisposable in my component

the dispose method is called when you press main refresh button on the browser 
in it I call another javascript method to remove my event listener

the javascript method is called but never completes using await JSRuntime.InvokeVoidAsync
any code after this call is not executed and my console.log in the dispose is not shown in the console (even if i add a log to the very start of the JS method)

after waiting for the garbage collector to fire it throws the error System.Threading.Tasks.TaskCanceledException A task was canceled, on my javascript call in the dispose method.

If I navigate to another page instead of pressing refresh all works correctly as expected.

I am using Microsoft edge but it happens in other browsers as well
Microsoft Visual Studio Professional 2019
Version 16.8.4
framework: .Net5.0

here is a cut down version of my code 

```
@implements IDisposable;
@inject IJSRuntime JSRuntime;

<input id="@this.ID" class="NumericalInput" />

[Parameter]
public string ID { get; set; }

protected async override void OnAfterRender(bool firstRender)
{
   if (firstRender == true)
   {
       await JSRuntime.InvokeVoidAsync("AddInputKeyboardEventHandler", this.ID);
   }
}

public async void Dispose()
{
   await JSRuntime.InvokeVoidAsync("RemoveInputKeyboardEventHandler", this.ID);
}
```

-----
javascript
-----
```
function AddInputKeyboardEventHandler(p_ControlID) {
   var objControl = document.getElementById(p_ControlID);
   if (objControl != null) {
     objControl.addEventListener("keydown", InputKeyboardEventHandler);
    console.log("AddInputKeyboardEventHandler");
  }
}

function RemoveInputKeyboardEventHandler(p_ControlID) {
   var objControl = document.getElementById(p_ControlID);
   if (objControl != null) {
      objControl.removeEventListener("keydown", InputKeyboardEventHandler);
      console.log("RemoveInputKeyboardEventHandler");
   }
}

function InputKeyboardEventHandler(e) {
   console.log("keydown " + e.keyCode + "   - " + e.shiftKey + " - " + this.id);
}
```

**Comment by javiercn** [comment]
@ScottyBlazor thanks for contacting us.

You should be implementing `IAsyncDisposable` and **never** use `async void` when you can avoid it.

**Comment by ScottyBlazor** [comment]
Hi Javier thanks for the quick reply

I have been doing a lot of reading on IAsyncDisposable but am still experiencing some issues mainly the "A task was canceled" error is still being thrown.

1. I have implemented the IAsyncDisposable and moved the javascript dispose call to it
```
@implements IAsyncDisposable;  
async ValueTask IAsyncDisposable.DisposeAsync()
{
   await JSRuntime.InvokeVoidAsync("RemoveInputKeyboardEventHandler", this.ID);
}
```
the IAsyncDisposable is called but it still throws "System.Threading.Tasks.TaskCanceledException A task was canceled" once the garbage collector fires.
If I put break points before and after the javascript call the code never reaches the break point after it.
also any console.log calls in the javascript RemoveInputKeyboardEventHandler are never printed to the browser console (might be cleared because of the page refresh), I added one to the start of the javascript method to try and confirm it is actually being called correctly and one after it.
If I navigate to a new page all my break points are hit correctly, and all my javascript console.log are printed, but when refreshing the page this is not the case.

2. I also have some local references in my component that need disposing, but implementing both the IDisposable and IAsyncDisposable stops the IDisposable.Dispose being called, even after removing the async keyword from it as you recommended.
I could add these synchronous disposes to the IAsyncDisposable but this page seems to suggest both should work
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync
"Implement both dispose and async dispose patterns"
 
3. Should I be adding the GC.SuppressFinalize(this); to my IAsyncDisposable.DisposeAsync and IDisposable.Dispose
My understanding of this is to let the garbage collector know we have manually cleaned up the object so it doesn't need to (but not quite sure). I have seen code samples where some people call it and some don't.

4. bit of a async noob question here. 
as my javascript RemoveInputKeyboardEventHandler does not return a value should I be awaiting it when calling InvokeVoidAsync.
or should I make it return a value or something just so it can be awaited correctly?

5. does my javascript code in RemoveInputKeyboardEventHandler need to be async, ie "async function RemoveInputKeyboardEventHandler(p_ControlID)"

Sorry to bombard you with questions, I just want to make my code work correctly without memory leaks and I find understanding the documentation is a little tricky in some of these areas.
